src/main/scala/pricemigrationengine/model/BillingDate.scala

```diff
@@ -0,0 +1,565 @@
+package pricemigrationengine.model
```

Comment:
As you say this is quite a chunky bit of code to copy over from support-service-lambdas.

I dont think we have 100% confidence in this code being able to predict the next bill date for every case as demonstrated by the alerts we get from  zuora-creditor.

Its probably close enough for this use case, although we could try using the zuora billing preview endpoint. In retrospect thats probably the route I would have gone down in support service lambdas.

If we bring this code across we should definitely bring over the test cases as well, ie a test for each of the different 'types' of subs ie:

https://github.com/guardian/support-service-lambdas/tree/master/lib/zuora-core/src/test/resources

It should just be a case of bringing over the tests from support-service-lambdas and adapting slightly.

Reply:
I'll bring the test cases in.  I'm going to add all the variants of the voucher cases as well.  So I'll do that in this PR.
The complexity of the billing date code is a bit alarming.  I'll add a card to investigate the billing preview too.